lua/ttt2/extensions/net.lua

```diff
@@ -6,58 +6,155 @@
+-- Can be up to 65.533KB see: https://wiki.facepunch.com/gmod/net.Start
 net.STREAM_FRAGMENTATION_SIZE = 65400
 
 -- Stream cache variables
-net.stream_cache = {}
+net.send_stream_cache  = {}
+net.receiving_players  = {}
+net.receive_stream_cache = {}
 net.stream_callbacks = {}
 
+--- 
+-- Sends next part of the stream
+-- @param string messageId a unique message id similar to the network strings
+-- @param number streamId the current stream number the split is requested for
+-- @param number split the part of the Stream that should be sent
+-- @realm shared
+-- @internal
+local function SendNextStream(messageId, streamId, split, plys)
+	net.Start(NETMSG_STREAM)
+	-- Write the messageId
+	net.WriteString(messageId)
+	-- Write the streamId
+	net.WriteUInt(streamId, 32)
+	-- Write the current split
+	net.WriteUInt(split, 8)
+
+	local data = net.send_stream_cache[messageId][streamId]
+	-- Write the actual data fragment as a string, which internally will also send its size
+	net.WriteString(data[split])
+
+	if SERVER then
+		if plys then
+			net.Send(plys)
+		else
+			net.Broadcast()
+		end
+	else
+		net.SendToServer()
+	end
+
+	if split <= 1 then
+		local eligiblePlayerList = net.receiving_players[messageId][streamId]
+
+		if eligiblePlayerList then
+			plys = istable(plys) and plys or {plys}
+
+			for i = 1, #plys do
+				eligiblePlayerList[plys[i]:SteamID64()] = nil
+			end
+
+			if next(eligiblePlayerList) ~= nil then return end
+
+			net.receiving_players[messageId][streamId] = nil
+		end
+
+		net.send_stream_cache[messageId][streamId] = nil
+	end
+end
+
+---
+-- It sends the next requested part of the stream and checks if the player is eligible for it
+-- @param number len Length of message in Bits
+-- @param Player ply player that the message has received, `nil` on the client
+-- @realm shared
+-- @internal
+local function SendNextSplit(len, ply)
+	local messageId = net.ReadString()
+	local streamId = net.ReadUInt(32)
+	local nextSplit = net.ReadUInt(8)
+
+	local eligiblePlayerList = net.receiving_players[messageId][streamId]
+
+	if nextSplit < 1 or eligiblePlayerList and not eligiblePlayerList[ply:SteamID64()] then return end
+
+	SendNextStream(messageId, streamId, nextSplit, ply)
+end
+net.Receive(NETMSG_REQUEST_NEXT_SPLIT, SendNextSplit)
+
 ---
 -- Initiates a stream message, usually for data that can be longer than
 -- the 64kb limit of a single net message. This will split up the data and send them in
 -- smaller fragments. The data will be converted (with sPON) to an encoded string during this process.
 --
 -- @param string messageId A unique message id similar to the network strings
--- @param table data The data table to send, this will be reconstructed at the client.
--- @param[opt] table|player client SERVERSIDE only! Optional, use it to send a stream to a single client or a group of clients.
+-- @param table data The data table to send, this will be reconstructed at the player.
```

Comment:
```suggestion
-- @param table data The data table to send, this will be reconstructed at the destination.
```

This could also be the server